### PR TITLE
P2415 What Is A `view`?

### LINKED ISSUE #2238: P2415R2 What Is A `view`?
[P2415R2](https://wg21.link/P2415R2) What Is A `view`?

Note: If this link appears to be broken, don't worry! Papers are often revised during a WG21 meeting before the final vote. When that happens, there's a delay of a few weeks between the end of the meeting and the publication of the papers. When the papers are available, the https://wg21.link redirector will start working.

This increases the value of `__cpp_lib_ranges`.

### PR DESCRIPTION
**Pull request by CaseyCarter**
Add the move-only range adaptor `owning_view`, and use it to implement `views::all` changes to correspond with `viewable_range` changes.

Resolves #2238

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -1379,11 +1382,9 @@ namespace ranges {
         return basic_istream_view<_Ty, _Elem, _Traits>{_Stream};
     }
 
-    // clang-format off
```
Comment (miscco) [inline stl/inc/ranges:1382]: I am a bit unsure about that one. Usually clang-format would bail here

File: stl/inc/ranges [hunk 2]
```diff
@@ -34,7 +34,10 @@ namespace ranges {
     template <class _Rng>
     concept viewable_range = range<_Rng>
         && ((view<remove_cvref_t<_Rng>> && constructible_from<remove_cvref_t<_Rng>, _Rng>)
-            || (!view<remove_cvref_t<_Rng>> && borrowed_range<_Rng>));
+            || (!view<remove_cvref_t<_Rng>>
+                && (is_lvalue_reference_v<_Rng>
+                    || (movable<remove_reference_t<_Rng>>
+                        && !_Is_specialization_v<remove_cvref_t<_Rng>, initializer_list>))));
```
Comment (miscco) [inline stl/inc/ranges:40]: The gift that never stops giving

File: stl/inc/ranges [hunk 3]
```diff
@@ -1435,20 +1436,101 @@ namespace ranges {
     template <class _Rng>
     inline constexpr bool enable_borrowed_range<ref_view<_Rng>> = true;
 
+    // clang-format off
+    template <range _Rng>
+        requires (movable<_Rng> && !_Is_specialization_v<remove_cvref_t<_Rng>, initializer_list>)
```
Comment (miscco) [inline stl/inc/ranges:1441]: I would prefer if we could put that into a named concept

File: stl/inc/ranges [hunk 4]
```diff
@@ -1435,20 +1436,101 @@ namespace ranges {
...
     namespace views {
         template <class _Rng>
         concept _Can_ref_view = requires(_Rng&& __r) {
             ref_view{static_cast<_Rng&&>(__r)};
         };
 
         template <class _Rng>
-        concept _Can_subrange = requires(_Rng&& __r) {
-            subrange{static_cast<_Rng&&>(__r)};
+        concept _Pwnable = requires(_Rng&& __r) {
```
Comment (miscco) [inline stl/inc/ranges:1527]: Should that be `_Ownable` ? 

That said, the typo might be worth keeping

Also can we change this to `_Is_ownable` and use `_Ownable` for the aforementioned helper concept?

File: tests/std/tests/P2415R2_owning_view/test.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                STATIC_ASSERT(
+                    is_nothrow_default_constructible_v<owning_view<R>> == is_nothrow_default_constructible_v<R>);
+                if constexpr (default_initializable<R>) {
+                    [[maybe_unused]] owning_view<R> default_constructed;
+                }
+
+                owning_view<R> value_constructed{R{input}};
+                STATIC_ASSERT(is_nothrow_constructible_v<owning_view<R>, R> == is_nothrow_move_constructible_v<R>);
+
+                auto move_constructed = move(value_constructed);
```
Comment (miscco) [inline tests/std/tests/P2415R2_owning_view/test.cpp:49]: can we make this `owning_view<R>` instead of `auto`

File: stl/inc/ranges (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/ranges:1382]: I don't know why clang-format wants to actually do the right thing here, but it does. This region is definitely "clang-format on" (from line 1312). I discovered this when I damaged the comments while copying `ref_view` to `owning_view`.

File: stl/inc/ranges [hunk 6]
```diff
@@ -3513,8 +3595,8 @@ namespace ranges {
                 const auto _Pat_end = _RANGES end(_It._Parent->_Pattern);
                 auto _Last          = _RANGES end(_It._Parent->_Range);
                 if constexpr (_Tiny_range<_Pat>) {
-                    const auto& _Cur = _It._Get_current(); // Intentionally a reference. Since _Pat is tiny, this could
-                                                           // be a move-only iterator type.
+                    const auto& _Cur = _It._Get_current(); // Intentionally a reference. Since _Pat is tiny, this
+                                                           // could be a move-only iterator type.
```
Comment (CaseyCarter) [inline stl/inc/ranges:3599]: This looks like damage from an inadvertent, reverted increase in indentation.

File: stl/inc/ranges (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/ranges:1441]: On first read I misinterpreted this as a request to put `_Is_specialization_v<remove_cvref_t<_Rng>, initializer_list>` into an `_Is_initializer_list` helper so this would more closely reflect the wording from the working draft. This makes a ton of sense to me.

Your comment on 1527, however, implies that you want the entire *requires-clause* (and presumably the *type-constraint* as well) pulled out into a helper concept. I'm inclined not to do so, again because I'd like the wording to more closely reflect the working draft. Do you have a motivating concern?

File: stl/inc/ranges (hunk 2 above)
Comment (miscco) [reply stl/inc/ranges:40]: Can we reuse `_Is_initializer_list` here too?

Comment (CaseyCarter) [reply stl/inc/ranges:40]: Good idea - I'll have to bring back the `remove_cvref_t`, which will again justify it being a distinct trait.

**Review by miscco — APPROVED** [review]
Grain of salt: the paper has not yet been available

**Comment by CaseyCarter** [comment]
> Grain of salt: the paper has not yet been available

[P2415R2.zip](https://github.com/microsoft/STL/files/7346155/P2415R2.zip) (Zipped because GitHub refuses `.html` attachments :shrug:)

File: stl/inc/ranges [hunk 7]
```diff
@@ -30,11 +30,17 @@ _STD_BEGIN
...
+
     // clang-format off
     template <class _Rng>
     concept viewable_range = range<_Rng>
         && ((view<remove_cvref_t<_Rng>> && constructible_from<remove_cvref_t<_Rng>, _Rng>)
-            || (!view<remove_cvref_t<_Rng>> && borrowed_range<_Rng>));
+            || (!view<remove_cvref_t<_Rng>>
+                && (is_lvalue_reference_v<_Rng>
+                    || (movable<remove_reference_t<_Rng>>
+                        && !_Is_initializer_list<_Rng>))));
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges:43]:
```suggestion
                    || (movable<remove_reference_t<_Rng>> && !_Is_initializer_list<_Rng>))));
```

File: tests/std/tests/P2415R2_owning_view/test.cpp [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+            { // constructors and assignment operators
+                STATIC_ASSERT(constructible_from<owning_view<R>, R>);
+                STATIC_ASSERT(!constructible_from<owning_view<R>, R&>);
+                STATIC_ASSERT(!constructible_from<owning_view<R>, const R&>);
+                STATIC_ASSERT(!constructible_from<owning_view<R>, const R>);
+
+                STATIC_ASSERT(default_initializable<owning_view<R>> == default_initializable<R>);
+                STATIC_ASSERT(
+                    is_nothrow_default_constructible_v<owning_view<R>> == is_nothrow_default_constructible_v<R>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P2415R2_owning_view/test.cpp:41]: Technically should ~~`import std;`~~ `#include <type_traits>`.

File: tests/std/tests/P2415R2_owning_view/test.cpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                if constexpr (default_initializable<R>) {
+                    [[maybe_unused]] owning_view<R> default_constructed;
+                }
+
+                owning_view<R> value_constructed{R{input}};
```
Comment (StephanTLavavej) [inline tests/std/tests/P2415R2_owning_view/test.cpp:46]: Naming nitpick: After a variable named `default_constructed`, this name `value_constructed` seems confusing, given the Standard terminology of value-initialized.

File: tests/std/tests/P0896R4_views_drop/test.cpp [hunk 10]
```diff
@@ -93,7 +88,7 @@ using pipeline_t = mapped_t<mapped_t<mapped_t<mapped_t<Rng>>>>;
 
 template <class Rng>
 concept CanViewDrop = requires(Rng&& r) {
-    views::drop(static_cast<Rng&&>(r), 42);
+    views::drop(std::forward<Rng>(r), 42);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_drop/test.cpp:91]: This test is already `using namespace std;`.

File: tests/std/tests/P0896R4_views_drop/test.cpp [hunk 11]
```diff
@@ -161,8 +156,8 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
     }
 
     // ... with rvalue argument
-    STATIC_ASSERT(CanViewDrop<remove_reference_t<Rng>> == is_view || enable_borrowed_range<remove_cvref_t<Rng>>);
-    if constexpr (is_span<remove_cvref_t<Rng>> || is_view) {
+    STATIC_ASSERT(CanViewDrop<remove_reference_t<Rng>> == is_view || movable<remove_reference_t<Rng>>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_drop/test.cpp:159]: In `P0896R4_views_all`, I saw you changed

```diff
-static_assert(CanViewAll<remove_cvref_t<Rng>> == is_view || ranges::enable_borrowed_range<remove_cvref_t<Rng>>);
+static_assert(CanViewAll<remove_reference_t<Rng>> == (is_view || movable<remove_reference_t<Rng>>) );
```

The updated lines in that file and this file look very similar, except for the parentheses on the RHS, which do affect the meaning (as equality has higher precedence than logical OR).

Is this intentional? `P0896R4_views_drop` appears to be the only file without such parentheses.

File: tests/std/tests/P0896R4_views_elements/test.cpp [hunk 12]
```diff
@@ -134,18 +132,7 @@ constexpr bool test_one(Rng&& rng) {
...
-
-        STATIC_ASSERT(same_as<decltype(views::elements<0>(move(as_const(rng)))), RS>);
-        STATIC_ASSERT(noexcept(views::elements<0>(move(as_const(rng)))) == is_noexcept);
-
-        STATIC_ASSERT(same_as<decltype(move(as_const(rng)) | views::elements<0>), RS>);
-        STATIC_ASSERT(noexcept(move(as_const(rng)) | views::elements<0>) == is_noexcept);
-
-        STATIC_ASSERT(same_as<decltype(move(as_const(rng)) | pipeline), pipeline_t<const remove_reference_t<Rng>>>);
-        STATIC_ASSERT(noexcept(move(as_const(rng)) | pipeline) == is_noexcept);
+        STATIC_ASSERT(always_false<Rng>); // FIXME
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_elements/test.cpp:135]: Fixme!

File: tests/std/tests/P0896R4_views_join/test.cpp [hunk 13]
```diff
@@ -34,98 +34,93 @@ constexpr bool test_one(Outer&& rng, Expected&& expected) {
...
 
+        using V = views::all_t<Outer>;
+        using R = join_view<V>;
+        static_assert(ranges::view<R>);
+        static_assert(input_range<R> == input_range<Inner>);
+        static_assert(forward_range<R> == (deref_is_glvalue && forward_range<Outer> && forward_range<Inner>) );
+        // clang-format off
+        static_assert(bidirectional_range<R> ==
+            (deref_is_glvalue && bidirectional_range<Outer> && bidirectional_range<Inner> && common_range<Inner>));
+        // clang-format on
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_join/test.cpp:120]: Can an empty comment be used here to force the line wrap?

File: tests/std/tests/P0896R4_views_reverse/test.cpp [hunk 14]
```diff
@@ -143,21 +141,7 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
...
-
-        static_assert(same_as<decltype(move(as_const(rng)) | views::reverse), RS>);
-        static_assert(noexcept(move(as_const(rng)) | views::reverse) == is_noexcept);
-
-        static_assert(same_as<decltype(move(as_const(rng)) | views::reverse | views::reverse | views::reverse), RS>);
-        static_assert(noexcept(move(as_const(rng)) | views::reverse | views::reverse | views::reverse) == is_noexcept);
-
-        static_assert(same_as<decltype(move(as_const(rng)) | pipeline), RS>);
-        static_assert(noexcept(move(as_const(rng)) | pipeline) == is_noexcept);
+        static_assert(always_false<Rng>); // FIXME
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_reverse/test.cpp:144]: Fixmeow!

File: stl/inc/ranges [hunk 15]
```diff
@@ -1454,20 +1457,101 @@ namespace ranges {
...
+            noexcept(noexcept(_RANGES end(_Range))) /* strengthened */ requires range<const _Rng> {
+            return _RANGES end(_Range);
+        }
+
+        _NODISCARD constexpr bool empty() noexcept(noexcept(_RANGES empty(_Range))) /* strengthened */
+            requires _Can_empty<_Rng> {
+            return _RANGES empty(_Range);
+        }
+        _NODISCARD constexpr bool empty() const noexcept(noexcept(_RANGES empty(_Range))) /* strengthened */
+            requires _Can_empty<const _Rng> {
```
Comment (barcharcraz) [inline stl/inc/ranges:1515]: Can't we write the requires requires constraint directly? or does that need requires expression support that we don't have?

File: tests/std/tests/P0896R4_views_take_while/test.cpp [hunk 16]
```diff
@@ -31,7 +31,7 @@ using pipeline_t = ranges::take_while_view<
 
 template <class Rng>
 concept CanViewTakeWhile = requires(Rng&& r) {
-    views::take_while(static_cast<Rng&&>(r), is_less_than<3>);
+    views::take_while(forward<Rng>(r), is_less_than<3>);
```
Comment (barcharcraz) [inline tests/std/tests/P0896R4_views_take_while/test.cpp:34]: why are we changing all these? I don't have a problem changing them but am curious.

**Review by barcharcraz — APPROVED** [review]

File: tests/std/tests/P0896R4_views_take_while/test.cpp (hunk 16 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_views_take_while/test.cpp:34]: Just general cleanup. Stephan feels that `static_cast<meow&&>(woof)` harms readability enough to not be worth the slight increase in throughput over `forward<meow>(woof)`. We reached an accord that allows the cast form to be used in constraints, since the cost of specializing `forward` for argument types with which it may actually never be called is slightly more impactful than only instantiating when we know it will be called. Some of the allowed uses have leaked out into contexts where they aren't allowed, and I clean them up when I notice them.

File: stl/inc/ranges (hunk 15 above)
Comment (CaseyCarter) [reply stl/inc/ranges:1515]: `view_interface` has four occurrences of this same ad-hoc constraint, so I made a helper concept for DRY purposes. @miscco implemented `ref_view` very shortly thereafter and reused the helper, I assume also just for DRY. I copy-pasted `ref_view` to make `owning_view` and didn't give this constraint any thought at all.

TLDR: Everybody hates `requires requires`. Shout if you'd prefer that we spell it out explicitly here (and/or elsewhere) so as to more obviously agree with the Standard wording. Both are equivalent under as-if.

Comment (StephanTLavavej) [reply stl/inc/ranges:1515]: I'm happy with the named concept.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
2 years after C++20 was published, I'm glad we've finally found the answer to what a `view` is! :joy_cat: :eye: :tada:
